scio-core/src/main/scala/com/spotify/scio/ScioContext.scala

```diff
@@ -239,6 +239,37 @@ object ContextAndArgs {
     }
   }
 
+  final case class PipelineOptionsParser[T <: PipelineOptions: ClassTag] private ()
+      extends ArgsParser[Try] {
+    override type ArgsType = T
+
+    override def parse(args: Array[String]): Try[Result] = Try {
+      val (opts, remainingArgs) = ScioContext.parseArguments[T](args, withValidation = true)
+      Either.cond(remainingArgs.asMap.isEmpty, (opts, opts), s"Unused $remainingArgs")
+    } match {
+      case r @ Success(Right(_)) => r
+      case Success(Left(v))      => Failure(new Exception(v))
+      case f                     => f
+    }
+  }
+
+  sealed trait TypedArgsParser[T, F[_]] {
+    def parser: ArgsParser[F]
+  }
+
+  sealed trait LowPrioTypedArgsParser {
+    implicit def caseApp[T: Parser: Help]: TypedArgsParser[T, Try] = new TypedArgsParser[T, Try] {
```

Comment:
Would a `T <: Product` here work as alternative to low-prio hack?

Reply:
Yeah good point, i think it will.

Reply:
I was thinking about case-app command support which is a sum type ... but i think it will not actually work with Parse/Help bounds